### ISSUE TITLE
Redesign LinkGenerationTemplate

### ISSUE THREAD

**Issue by rynowak** [aspnet/Routing#866]

**Comment by drieseng** [comment]
I had a quick look at **DefaultLinkGenerator** and **DefaultLinkGenerationTemplate** before I saw this issue. What I noticed was that:
* it always iterates over all endpoints (it does this for the purpose of logging only?)
* for each endpoint it has to check whether it's a **RouteEndpoint**
* it then creates a **List\<RouteEndpoint>** containing all **RouteEndpoint** instances.

It does this for each call to `GetPathByAddress(…)`, `GetUriByAddress(…)` and `GetTemplateByAddress(…)`.

Can you give me a little more background:
* How many endpoints are typically returned by `IEndpointAddressScheme.FindEndpoints(TAddress)` in an average application?
* It is necessary to always log all **RouteEndpoint** instances (which rules out only enumerating the endpoints until a match)?
* How often is `DefaultLinkGenerator.GetTemplateByAddress(…)` invoked in the lifetime of a typical application (whatever that is :p), and how does this compare to the methods on  **DefaultLinkGenerator** that do not take a **List\<RouteEndpoint>** as argument.
* Are these classes used often enough to warrant "tuning"?

Would it be better to use a different link generator in function of the address scheme?
For example, in case of **EndpointNameAddressScheme** we know there's either one or no matching endpoint. No need to allocate and iterate through a list of endpoints.

**Comment by drieseng** [comment]
@JamesNK @rynowak Can you feedback on my comment above? Is this something I can take a stab at?

**Comment by drieseng** [comment]
> Where is this? I see code that's iterating but it logs each endpoint as it goes, it doesn't iterate all of them and log up front. 

in [`GetEndpoints<TAddress>(TAddress address)`](https://github.com/aspnet/Routing/blob/3d828221a19d91907c52c2f40928b019bee1ef92/src/Microsoft.AspNetCore.Routing/DefaultLinkGenerator.cs#L192) which is invoked from just about any method in this class that does not take a **List\<RouteEndpoint>**.

> Yes, but micro-optimizations like avoiding a loop when we have a single endpoint aren't really relevant compared to the cost of actually processing a template.

If such a component is heavily used, then any optimization (whether perf or memory wise) makes sense to me. I'm not familiar enough with the code nor its usage (professional I mostly work on back end stuff) to perform a huge refactoring, but I'm willing to help out wherever possible. Such micro-optimizations are for me a way to familiarize myself with the code, and at the same time provide a (micro)benefit to the project.

Thanks for the feedback!

**Comment by rynowak** [comment]
> in GetEndpoints<TAddress>(TAddress address) which is invoked from just about any method in this class that does not take a List<RouteEndpoint>.

Ah, you're right! That is something we should clean up.

**Comment by drieseng** [comment]
I'd be happy to take care of it. Do you want to continue logging all endpoints that are found (when debug logging is enabled)? If yes, we may need to have fast and slow paths. In the former case we'd only iterate until we find a match.

I had a quick look at **TemplateBinder**, and identified some low hanging fruit.
I'll look at these first.

**Comment by rynowak** [comment]
I think it would be OK to log each endpoint as we try it in Debug.

Feel free to send a PR for these things 👍

**Comment by drieseng** [comment]
In what repo can I find **Microsoft.Extensions.Internal.PropertyHelper**?
It's included through a (source) package reference in the **Microsoft.AspNetCore.Routing.Abstractions** project.

**Comment by rynowak** [comment]
https://github.com/aspnet/Extensions/tree/master/src/Shared/src/PropertyHelper

**Comment by rynowak** [comment]
Yeah please send a PR for this. If this change is what you say then we absolutely want it improve it.
